File: BaekjoonOnlineJudge/2805/2805.py
import sys

input = sys.stdin.readline


def get_tree(cut_height):
    cut_tree = 0

    for tree in trees:
        if tree > cut_height:
            cut_tree += tree - cut_height
    
    return cut_tree


N, M = map(int, input().split())
trees = list(map(int, input().split()))

answer = 0
left = 0
right = max(trees)

while left <= right:
    mid = (left + right) // 2

    if M > get_tree(mid):        
        right = mid - 1
    else:
        answer = max(answer, mid)
        left = mid + 1

print(answer)

File: BaekjoonOnlineJudge/2805/2805.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int n, m;
    private static int[] trees;

    private static long getTree(int height) {
        return Arrays.stream(trees).filter(tree -> tree > height).mapToLong(tree -> tree - height)
            .sum();
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        trees = new int[n];

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++) {
            trees[i] = Integer.parseInt(st.nextToken());
        }

        int answer = 0;
        int left = 0;
        int right = Arrays.stream(trees).max().getAsInt();

        while (left <= right) {
            int mid = (left + right) / 2;

            long cutTree = getTree(mid);

            if (cutTree < m) {
                right = mid - 1;
            } else {
                answer = Math.max(answer, mid);
                left = mid + 1;
            }
        }

        bw.write(String.valueOf(answer));

        bw.flush();
        bw.close();
        br.close();
    }
}